gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/CommitActivityImpl.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.gobblin.util.Either;
+import org.apache.gobblin.util.ExecutorsUtils;
+import org.apache.gobblin.util.executors.IteratorExecutor;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+
+@Slf4j
+public class CommitActivityImpl implements CommitActivity {
+
+  static int DEFAULT_NUM_DESERIALIZATION_THREADS = 10;
+  static int DEFAULT_NUM_COMMIT_THREADS = 1;
+  @Override
+  public int commit(WUProcessingSpec workSpec) {
+    // TODO: Make this configurable
+    int numDeserializationThreads = DEFAULT_NUM_DESERIALIZATION_THREADS;
+    try {
+      FileSystem fs = Help.loadFileSystem(workSpec);
+      JobState jobState = Help.loadJobState(workSpec, fs);
+      SharedResourcesBroker<GobblinScopeTypes> instanceBroker = JobStateUtils.getSharedResourcesBroker(jobState);
+      JobContext globalGobblinContext = new JobContext(jobState.getProperties(), log, instanceBroker, null);
+      // TODO: Task state dir is a stub with the assumption it is always colocated with the workunits dir (as in the case of MR which generates workunits)
+      Path jobIdParent = new Path(workSpec.getWorkUnitsDir()).getParent();
+      Path jobOutputPath = new Path(new Path(jobIdParent, "output"), jobIdParent.getName());
+      log.info("Output path at: " + jobOutputPath + " with fs at " + fs.getUri());
+      StateStore<TaskState> taskStateStore = Help.openTaskStateStore(workSpec, fs);
+      Collection<TaskState> taskStateQueue =
+          ImmutableList.copyOf(
+              TaskStateCollectorService.deserializeTaskStatesFromFolder(taskStateStore, jobOutputPath, numDeserializationThreads));
+      commitTaskStates(jobState, taskStateQueue, globalGobblinContext);
+      return taskStateQueue.size();
+    } catch (Exception e) {
+      //TODO: IMPROVE GRANULARITY OF RETRIES
+      throw ApplicationFailure.newNonRetryableFailureWithCause(
+          "Failed to commit dataset state for some dataset(s) of job <jobStub>",
+          IOException.class.toString(),
+          new IOException(e),
+          null
+      );
+    }
+  }
+
+  /**
+   * Commit task states to the dataset state store.
+   * @param jobState
+   * @param taskStates
+   * @param jobContext
+   * @throws IOException
+   */
+  void commitTaskStates(State jobState, Collection<TaskState> taskStates, JobContext jobContext) throws IOException {
+    Map<String, JobState.DatasetState> datasetStatesByUrns = createDatasetStatesByUrns(taskStates);
+    final boolean shouldCommitDataInJob = JobContext.shouldCommitDataInJob(jobState);
+    final DeliverySemantics deliverySemantics = DeliverySemantics.AT_LEAST_ONCE;
+    //TODO: Make this configurable
+    final int numCommitThreads = DEFAULT_NUM_COMMIT_THREADS;
+    if (!shouldCommitDataInJob) {
+      log.info("Job will not commit data since data are committed by tasks.");
+    }
+
+    try {
+      if (!datasetStatesByUrns.isEmpty()) {
+        log.info("Persisting {} dataset urns.", datasetStatesByUrns.size());
```

Comment:
let's add info about the dataset otherwise this cannot be connected to a particular job